src/service-override/views.ts

```diff
@@ -197,7 +203,21 @@ type Label = string | {
 interface EditorPanelOption {
   readonly id: string
   name: string
-  renderBody (container: HTMLElement): IDisposable
+  renderBody (container: HTMLElement): BodyRenderer
```

Comment:
wouldn't it be more simple to have the `setInput` method next to the `renderBody` method?

Reply:
is it possible that `renderBody` is called without `setInput`? Can't we just provide the input as a parameter of `renderBody` instead of having a setInput method?

Reply:
> is it possible that renderBody is called without setInput? Can't we just provide the input as a parameter of renderBody instead of having a setInput method?

Unfortunately this is not possible. VSCode first creates the editor, and a couple ticks later it will call `setInput`. So it's not synchronously available. This was my first attempt, but after this I realised that the same editor will be reused between inputs.

Reply:
What about making the input parameter optional? (at the first render, it's undefined, then it's called every time the input change with the new input)

Reply:
Interesting... So I think for us that would be less interesting, because we render a React component for the editor, and then when the input changes we don't rerender the whole editor, but we change the prop of the component itself.

Hmm, though we _can_ check if it's already rendered, and then update the props if it's there. I'll take a look.

Reply:
I'm interested on how you implemented it with React, that was my concern, how do you update your existing component prop in the callback?

Reply:
I'm using portals, which allows me to render React components in the given element by `onRender`, while still in the same React context.

Code will say a thousand words:

```tsx
export const CustomEditor = ({
  globPattern,
  id,
  name,
  Component,
  getName,
}: CustomEditorProps) => {
  const [containers, setContainers] = React.useState<Container[]>([]);
  const { monacoEditor } = useEnvironmentInterface();
  const lastId = React.useRef(0);

  React.useEffect(() => {
    const disposablePromise = monacoEditor.registerEditor({
      id,
      name,
      globPattern,
      onRender: (element) => {
        const id = lastId.current++;
        setContainers((containers) => [
          ...containers,
          {
            id,
            element,
          },
        ]);

        return {
          dispose: () => {},
          setInput: async (editorInput) => {
            if (
              "setName" in editorInput &&
              typeof editorInput.setName === "function" &&
              getName
            ) {
              // This is a hacky way to set the title to the filename, but it's the
              // best way right now to set it. Ideally we use `EditorResourceAccessor`.

              editorInput.setName(getName(editorInput.resource));
            }

            setContainers((containers) =>
              containers.map((container) =>
                container.id === id
                  ? { ...container, filePath: editorInput.resource }
                  : container
              )
            );
          },
        };
      },
    });

    return () => {
      disposablePromise.then((disposable) => disposable.dispose());
    };
  }, []);

  return (
    <>
      {containers.map((container) =>
        createPortal(
          <Box
            css={{
              display: "block",
              width: "100%",
              height: "100%",
              padding: "$2",

              boxSizing: "border-box !important",
              "*": {
                boxSizing: "border-box !important",
              },
            }}
          >
            <Component key={container.id} resource={container.filePath} />
          </Box>,
          container.element
        )
      )}
    </>
  );
};
```

Reply:
Actually this reminds me that I should still implement dispose.

Reply:
That renders this:

<img width="1607" alt="image" src="https://github.com/CodinGame/monaco-vscode-api/assets/587016/0bd11f99-7465-44d2-9bfe-f2c33297f61d">

How I use the API:

```tsx
      <CustomEditor
        Component={({ resource }) => {
          if (!resource) {
            return <div>Loading...</div>;
          }

          const relativePath =
            pitcher.clients.fs.absoluteToRelativeWorkspacePath(
              resource.path as AbsoluteWorkspacePath
            );

          return (
            <div style={{ width: "100%", height: "100%" }}>
              <FileMarkdownPreview filepath={relativePath} />
            </div>
          );
        }}
        getName={(resource) => {
          if (resource) {
            return resource.path.split("/").pop() ?? "";
          }

          return "Markdown";
        }}
        globPattern="*.md"
        id="codesandbox.editors.markdown"
        name="Interactive Markdown"
      />
```

Reply:
Thank you, interesting :)